rustworkx-core/src/err.rs

```diff
@@ -0,0 +1,66 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+//! This module contains common error types and trait impls.
+
+use std::error::Error;
+use std::fmt::{Debug, Display, Formatter};
+
+#[derive(Debug)]
+pub enum ContractError {
+    DAGWouldCycle,
+}
+
+impl Display for ContractError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ContractError::DAGWouldCycle => fmt_dag_would_cycle(f),
+        }
+    }
+}
+
+impl Error for ContractError {}
+
+#[derive(Debug)]
+pub enum ContractSimpleError<E: Error> {
+    DAGWouldCycle,
+    MergeError(E),
+}
+
+impl<E: Error> Display for ContractSimpleError<E> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ContractSimpleError::DAGWouldCycle => fmt_dag_would_cycle(f),
+            ContractSimpleError::MergeError(ref e) => fmt_merge_error(f, e),
+        }
+    }
+}
+
+impl<E: Error> Error for ContractSimpleError<E> {}
+
+fn fmt_dag_would_cycle(f: &mut Formatter<'_>) -> std::fmt::Result {
+    write!(f, "The operation would introduce a cycle.")
+}
+
+fn fmt_merge_error<E: Error>(f: &mut Formatter<'_>, inner: &E) -> std::fmt::Result {
+    write!(f, "The prov failed with: {:?}", inner)
+}
+
+#[allow(dead_code)]
+fn fmt_node_not_found<N: Debug>(f: &mut Formatter<'_>, node: &N) -> std::fmt::Result {
+    write!(f, "Node not found in graph: {:?}", node)
+}
+
+#[allow(dead_code)]
+fn fmt_edge_not_found<E: Debug>(f: &mut Formatter<'_>, edge: &E) -> std::fmt::Result {
+    write!(f, "Edge not found in graph: {:?}", edge)
+}
```

Comment:
They're here mostly because I'm hoping to establish a pattern, and I'm not confident there's enough here otherwise for it to stick.

I can remove if it's your preference, let me know!

Reply:
Personally I'd rather remove the dead code or leave it as a comment or something. To me it's more confusing to leave this around I didn't see the pattern. I'm thinking if you want to make this a pattern I'd put this as a comment at the top using these as an example and explaining how they should be used.